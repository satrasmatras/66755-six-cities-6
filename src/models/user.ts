interface User {
  email: string,
  password: string
}

interface AuthInfo {
  avatarUrl: string,
  email: string,
  id: number,
  isPro: boolean,
  name: string
}

export default User;
