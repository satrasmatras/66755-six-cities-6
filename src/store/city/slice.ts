import {CITIES} from "../../mocks/cities";
import {createSlice} from "@reduxjs/toolkit";
import City from "../../models/city";

export interface CityState {
  city: City,
}

const initialState: CityState = {
  city: CITIES[0],
};

const citySlice = createSlice({
  name: `city`,
  initialState: initialState,
  reducers: {
    changeCity: (state, action) => {
      state.city = action.payload;
    }
  }
});

export const {
  changeCity,
} = citySlice.actions;

export default citySlice.reducer;
