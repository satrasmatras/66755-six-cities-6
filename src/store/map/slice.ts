import {createAction, createReducer} from "@reduxjs/toolkit";
import Offer from "../../models/offer";

export interface MapState {
  hoveredOffer: Offer
}

export const initialState: MapState = {
  hoveredOffer: null
};

export const UPDATE_HOVERED_OFFER = `map/updateHoveredOffer`;
export const updateHoveredOffer = createAction<Offer>(UPDATE_HOVERED_OFFER);

const mapReducer = createReducer(
    initialState,
    (builder) => {
      builder.addCase(updateHoveredOffer, ((state, action) => {
        state.hoveredOffer = action.payload;
      }));
    }
);

export default mapReducer;
