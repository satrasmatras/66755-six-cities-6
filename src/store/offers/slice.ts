import SortType from "../../models/sort-type";
import Offer from "../../models/offer";
import {createSlice} from "@reduxjs/toolkit";
import {ThunkDispatch} from "redux-thunk";
import {Action} from "redux";
import {RootState} from "../index";
import {AxiosInstance} from "axios";
import {adaptDataToOffer} from "../../adapters/offers";

export interface OffersState {
  offers: Offer[],
  sortType: SortType,
  isLoading: boolean,
}

const initialState: OffersState = {
  offers: [],
  sortType: SortType.POPULAR,
  isLoading: false,
};

const offersSlice = createSlice({
  name: `offers`,
  initialState,
  reducers: {
    updateOffers: (state, action) => {
      state.offers = action.payload;
    },
    setSortType: (state, action) => {
      state.sortType = action.payload;
    },
    setIsLoading: (state, action) => {
      state.isLoading = action.payload;
    },
  }
})

export const {
  updateOffers,
  setSortType,
  setIsLoading,
} = offersSlice.actions;

enum ApiRoutes {
  FETCH_OFFERS = `hotels`,
}

export const fetchOffers = () => (next: ThunkDispatch<undefined, undefined, Action>, _: RootState, api: AxiosInstance): void => {
  next(setIsLoading(true));
  api.get(ApiRoutes.FETCH_OFFERS)
    .then(({data}) => {
      next(updateOffers(data.map(adaptDataToOffer)));
      next(setIsLoading(false));
    });
};

export default offersSlice.reducer;
