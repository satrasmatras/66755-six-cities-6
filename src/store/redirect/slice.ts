import browserHistory from "../../services/browser-history";
import {Dispatch, Store} from "redux";
import {createAction} from "@reduxjs/toolkit";

export const REDIRECT_TO_ROUTE = `reducer/redirectToRoute`;
export const redirectToRoute = createAction<string>(REDIRECT_TO_ROUTE);

export const redirectMiddleware = (_store: Store) => (next: Dispatch) => (action: any) => {
  if (action.type === redirectToRoute.type) {
    browserHistory.push(action.payload);
  }

  return next(action);
};
