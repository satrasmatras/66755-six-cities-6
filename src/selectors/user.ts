import {RootState} from "../store";
import {createSelector} from "reselect";
import {AuthorizationStatus} from "../store/user/slice";

const selectAuthorizationStatus = (state: RootState) => state.user.authorizationStatus;

export const selectIsAuthorized = createSelector(
  selectAuthorizationStatus,
  (status) => status === AuthorizationStatus.AUTH
);
