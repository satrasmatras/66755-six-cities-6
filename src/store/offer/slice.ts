import Offer from "../../models/offer";
import Comment from "../../models/comment";
import {Dispatch} from "redux";
import {AxiosInstance} from "axios";
import {RootState} from "../index";
import {adaptDataToOffer} from "../../adapters/offers";
import {adaptDataToComment} from "../../adapters/comments";
import {CommentPost} from "../../models/comment-post";
import {createSlice} from "@reduxjs/toolkit";

interface OfferState {
  offer: Offer,
  comments: Comment[],
  offerIsLoading: false,
  commentsAreLoading: false,
}

const initialState: OfferState = {
  offer: null,
  comments: [],
  offerIsLoading: false,
  commentsAreLoading: false
};

const slice = createSlice({
  name: 'offer',
  initialState: initialState,
  reducers: {
    setOffer: (state, action) => {
      state.offer = action.payload;
    },
    setComments: (state, action) => {
      state.comments = action.payload;
    },
    addComment: (state, action) => {
      state.comments.push(action.payload);
    },
    setOfferIsLoading: (state, action) => {
      state.offerIsLoading = action.payload;
    },
    setCommentsAreLoading: (state, action) => {
      state.commentsAreLoading = action.payload;
    },
  }
})

export const {
  setOffer,
  setComments,
  setCommentsAreLoading,
  setOfferIsLoading,
  addComment
} = slice.actions;

const getOfferRoute = (id: number) => `/hotels/${id}`;

export const loadOfferById = (id: number) => async (dispatch: Dispatch, _: RootState, api: AxiosInstance) => {
  dispatch(setOfferIsLoading(true));
  const response = await api.get(getOfferRoute(id));
  const offer = response.data;
  dispatch(setOffer(adaptDataToOffer(offer)));
  dispatch(setOfferIsLoading(false));
}

const getCommentsRoute = (id: number) => `/comments/${id}`;

export const loadOfferComments = (id: number) => async (dispatch: Dispatch, _: RootState, api: AxiosInstance) => {
  dispatch(setCommentsAreLoading(true));
  const response = await api.get(getCommentsRoute(id));
  const comments =
    response
      .data
      .map(adaptDataToComment);
  dispatch(setComments(comments));
  dispatch(setCommentsAreLoading(false));
}

export const postComment = (commentPost: CommentPost, id: number) => async (dispatch: Dispatch, _: RootState, api: AxiosInstance) => {
  dispatch(setCommentsAreLoading(true));
  const response = await api.post(getCommentsRoute(id), commentPost);
  const comments =     response
    .data
    .map(adaptDataToComment);
  dispatch(setComments(comments));
  dispatch(setCommentsAreLoading(false));
}

export default slice.reducer;
