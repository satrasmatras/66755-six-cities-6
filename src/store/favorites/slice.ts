import Offer from "../../models/offer";
import {createSlice} from "@reduxjs/toolkit";
import {ThunkDispatch} from "redux-thunk";
import {Action} from "redux";
import {RootState} from "../index";
import {AxiosInstance} from "axios";
import {adaptDataToOffer} from "../../adapters/offers";
import {updateOffers} from "../offers/slice";
import {deleteItem} from "../../services/items";
import {setOffer} from "../offer/slice";

interface FavoritesState {
  favorites: Offer[],
  isLoading: boolean,
}

const initialState: FavoritesState = {
  favorites: [],
  isLoading: false,
};

const favoritesSlice = createSlice({
  name: `favorites`,
  initialState,
  reducers: {
    setFavorites: (state, action) => {
      state.favorites = action.payload;
    },
    deleteFavorite: (state, action) => {
      state.favorites = deleteItem(state.favorites, action.payload);
    },
    setIsLoading: ((state, action) => {
      state.isLoading = action.payload;
    })
  }
});

export const {
  setFavorites,
  deleteFavorite,
  setIsLoading
} = favoritesSlice.actions;

enum ApiRoutes {
  FAVORITES= `/favorite`,
}

export const fetchFavorites = () => async (next: ThunkDispatch<undefined, undefined, Action>, _: RootState, api: AxiosInstance): Promise<void> => {
  next(setIsLoading(true));
  const response = await api.get(ApiRoutes.FAVORITES);
  const favorites = response.data.map(adaptDataToOffer);
  next(setFavorites(favorites));
  next(setIsLoading(false));
};

export enum ToggleFavoriteTarget {
  MAIN,
  FAVORITES,
  OFFER,
}

const getFavoriteToggle = (id: number, value: number) => `/favorite/${id}/${value}`;

export const toggleFavorite = (offer: Offer, target: ToggleFavoriteTarget) => async (next: ThunkDispatch<undefined, undefined, Action>, _: RootState, api: AxiosInstance): Promise<void> => {
  const response = await api.post(getFavoriteToggle(offer.id, Number(!offer.isFavorite)));
  const updated = adaptDataToOffer(response.data);
  switch (target) {
    case ToggleFavoriteTarget.FAVORITES:
      next(deleteFavorite(updated));
      break;
    case ToggleFavoriteTarget.MAIN:
      next(updateOffers(updated));
      break;
    case ToggleFavoriteTarget.OFFER:
      next(setOffer(updated));
      break;
  }
};

export default favoritesSlice.reducer;
