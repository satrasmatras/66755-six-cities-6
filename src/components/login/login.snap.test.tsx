import React from "react";
import {render} from "@testing-library/react";
import {Router} from "react-router-dom";
import browserHistory from "../../services/browser-history";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import {initialState as userInitialState} from "../../store/user/slice";
import thunk from "redux-thunk";
import Login from "./login";

const mockStore = configureStore([thunk]);
const store = mockStore({
  user: userInitialState,
});

it(`Should cities list render correctly`, () => {
  const {container} = render(
      <Provider store={store}>
        <Router history={browserHistory}>
          <Login />
        </Router>
      </Provider>
  );
  expect(container).toMatchSnapshot();
});
