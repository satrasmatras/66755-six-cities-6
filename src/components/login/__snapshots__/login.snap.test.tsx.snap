// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should cities list render correctly 1`] = `
<div>
  <div
    style="display: none;"
  >
    <svg
      xmlns="http://www.w3.org/2000/svg"
    >
      <symbol
        id="icon-arrow-select"
        viewBox="0 0 7 4"
      >
        <path
          clip-rule="evenodd"
          d="M0 0l3.5 2.813L7 0v1.084L3.5 4 0 1.084V0z"
          fill-rule="evenodd"
        />
      </symbol>
      <symbol
        id="icon-bookmark"
        viewBox="0 0 17 18"
      >
        <path
          d="M3.993 2.185l.017-.092V2c0-.554.449-1 .99-1h10c.522 0 .957.41.997.923l-2.736 14.59-4.814-2.407-.39-.195-.408.153L1.31 16.44 3.993 2.185z"
        />
      </symbol>
      <symbol
        id="icon-star"
        viewBox="0 0 13 12"
      >
        <path
          clip-rule="evenodd"
          d="M6.5 9.644L10.517 12 9.451 7.56 13 4.573l-4.674-.386L6.5 0 4.673 4.187 0 4.573 3.549 7.56 2.483 12 6.5 9.644z"
          fill-rule="evenodd"
        />
      </symbol>
    </svg>
  </div>
  <div
    class="page page--gray page--login"
  >
    <header
      class="header"
    >
      <div
        class="container"
      >
        <div
          class="header__wrapper"
        >
          <div
            class="header__left"
          >
            <a
              class="header__logo-link"
              href="/"
            >
              <img
                alt="6 cities logo"
                class="header__logo"
                height="41"
                src="img/logo.svg"
                width="81"
              />
            </a>
          </div>
          <nav
            class="header__nav"
          >
            <ul
              class="header__nav-list"
            >
              <li
                class="header__nav-item user"
              >
                <a
                  class="header__nav-link header__nav-link--profile"
                  href="#"
                >
                  <div
                    class="header__avatar-wrapper user__avatar-wrapper"
                  />
                  <span
                    class="header__login"
                  >
                    Sign in
                  </span>
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </header>
    <main
      class="page__main page__main--login"
    >
      <div
        class="page__login-container container"
      >
        <section
          class="login"
        >
          <h1
            class="login__title"
            data-testid="login-title"
          >
            Sign in
          </h1>
          <form
            action="#"
            class="login__form form"
            method="post"
          >
            <div
              class="login__input-wrapper form__input-wrapper"
            >
              <label
                class="visually-hidden"
              >
                E-mail
              </label>
              <input
                class="login__input form__input"
                name="email"
                placeholder="Email"
                required=""
                type="email"
                value=""
              />
            </div>
            <div
              class="login__input-wrapper form__input-wrapper"
            >
              <label
                class="visually-hidden"
              >
                Password
              </label>
              <input
                class="login__input form__input"
                name="password"
                placeholder="Password"
                required=""
                type="password"
                value=""
              />
            </div>
            <button
              class="login__submit form__submit button"
              type="submit"
            >
              Sign in
            </button>
          </form>
        </section>
        <section
          class="locations locations--login locations--current"
        >
          <div
            class="locations__item"
          >
            <a
              class="locations__item-link"
              href="#"
            >
              <span>
                Amsterdam
              </span>
            </a>
          </div>
        </section>
      </div>
    </main>
  </div>
</div>
`;
